spec/datadog/tracing/contrib/concurrent_ruby/integration_test_spec.rb

```diff
@@ -58,40 +37,166 @@
     end
   end
 
-  describe 'patching' do
-    subject(:patch) do
-      Datadog.configure do |c|
-        c.tracing.instrument :concurrent_ruby
+  context 'Concurrent::Promises::Future' do
+    before(:context) do
+      # Execute an async future to force the eager creation of internal
+      # global threads that are never closed.
+      #
+      # This allows us to separate internal concurrent-ruby threads
+      # from ddtrace threads for leak detection. We need to create the maximum
+      # number of threads that will be created concurrently in a test, which in
+      # this case is 2.
+      ThreadHelpers.with_leaky_thread_creation(:concurrent_ruby) do
+        Concurrent::Promises.future {
+          Concurrent::Promises.future {}.value
+        }.value
       end
     end
 
-    it 'adds FuturePatch to Future ancestors' do
-      expect { patch }.to change { ::Concurrent::Future.ancestors.map(&:to_s) }
-        .to include('Datadog::Tracing::Contrib::ConcurrentRuby::FuturePatch')
+    subject(:deferred_execution) do
+      outer_span = tracer.trace('outer_span')
+      future = Concurrent::Promises.future do
+        tracer.trace('inner_span') {}
+      end
+
+      future.wait
+      outer_span.finish
     end
-  end
 
-  context 'when context propagation is disabled' do
-    it_behaves_like 'deferred execution'
+    describe 'patching' do
+      subject(:patch) do
+        Datadog.configure do |c|
+          c.tracing.instrument :concurrent_ruby
+        end
+      end
 
-    it 'inner span should not have parent' do
-      deferred_execution
-      expect(inner_span).to be_root_span
+      it 'adds PromisesFuturePatch to Promises ancestors' do
+        expect { patch }.to change { ::Concurrent::Promises.singleton_class.ancestors.map(&:to_s) }
+          .to include('Datadog::Tracing::Contrib::ConcurrentRuby::PromisesFuturePatch')
+      end
+    end
+
+    context 'when context propagation is disabled' do
+      it_behaves_like 'deferred execution'
+
+      it 'inner span should not have parent' do
+        deferred_execution
+        expect(inner_span).to be_root_span
+      end
+    end
+
+    context 'when context propagation is enabled' do
+      before do
+        Datadog.configure do |c|
+          c.tracing.instrument :concurrent_ruby
+        end
+      end
+
+      it_behaves_like 'deferred execution'
+
+      it 'inner span parent should be included in outer span' do
+        deferred_execution
+        expect(inner_span.parent_id).to eq(outer_span.span_id)
+      end
+
+      context 'when there are multiple futures with inner spans that have the same parent' do
```

Comment:
I added this new context to the PR at @marcotc's suggestion. This test fails with the previous implementation and passes when `continue_trace!` is used, confirming Marco's theory [here.](https://github.com/DataDog/dd-trace-rb/pull/1522#discussion_r1360666393)